crates/driver/src/domain/competition/auction.rs

```diff
@@ -11,25 +12,84 @@ use {
 /// solving them.
 #[derive(Debug)]
 pub struct Auction {
-    /// [`None`] if this auction applies to a quote.
+    /// [`None`] if this auction applies to a quote. See
+    /// [`crate::domain::quote`].
     pub id: Option<Id>,
-    pub tokens: Vec<Token>,
-    pub orders: Vec<competition::Order>,
+    pub orders: Orders,
+    pub tokens: Tokens,
     pub gas_price: eth::GasPrice,
     pub deadline: Deadline,
 }
 
-impl Auction {
-    pub fn is_trusted(&self, token: eth::TokenAddress) -> bool {
-        self.tokens
-            .iter()
-            .find(|t| t.address == token)
-            .map(|token| token.trusted)
-            .unwrap_or(false)
+/// The orders for an auction. The orders get sorted such that those which are
+/// more likely to be fulfilled come before less likely orders.
+#[derive(Debug)]
+pub struct Orders(Vec<competition::Order>);
+
+impl Orders {
+    pub fn new(mut orders: Vec<competition::Order>, tokens: &Tokens) -> Self {
+        // Sort orders such that most likely to be fulfilled come first.
+        orders.sort_by(|left, right| {
+            // Market orders are preferred over limit orders, as the expectation is that
+            // they should be immediately fulfillable. Liquidity orders come last, as they
+            // are the most niche and rarely used.
+            let left_kind = match left.kind {
+                competition::order::Kind::Market => 2,
+                competition::order::Kind::Limit { .. } => 1,
+                competition::order::Kind::Liquidity => 0,
+            };
+            let right_kind = match right.kind {
+                competition::order::Kind::Market => 2,
+                competition::order::Kind::Limit { .. } => 1,
+                competition::order::Kind::Liquidity => 0,
+            };
+            match left_kind.cmp(&right_kind) {
+                std::cmp::Ordering::Equal => {
+                    // If the orders are of the same kind, then sort by likelihood of fulfillment
+                    // based on token prices.
+                    left.likelihood(tokens).cmp(&right.likelihood(tokens))
+                }
+                other => other,
+            }
+        });
+        orders.reverse();
```

Comment:
I think this should be equivalent.
```suggestion
        orders.sort_by_key(|order| {
            // Market orders are preferred over limit orders, as the expectation is that
            // they should be immediately fulfillable. Liquidity orders come last, as they
            // are the most niche and rarely used.
            let class = match order.kind {
                competition::order::Kind::Market => 2,
                competition::order::Kind::Limit { .. } => 1,
                competition::order::Kind::Liquidity => 0,
            };
            std::cmp::Reverse((class, order.likelihood(tokens)))
        });
```

Reply:
https://github.com/cowprotocol/services/pull/1678/commits/32c5a55cdf0695e5ef9293fee9f4206d14755443